### PR TITLE
Cleanup old oplfex devices

### PR DESCRIPTION
**Pull request by akhilamohanan**
* Check if there are any other opflex-odevs in the list which have a
different “fabricPathDn” than the last one
* Flag these entries for delete after a configurable period which defaults to 30 minutes
* Just before finally deleting the entry check if the opflex-odev list has changed and if the opflex-odevs we are trying to delete are not the only ones in the list or are currently being used

### CODE REVIEW

File: pkg/controller/services.go [hunk 1]
```diff
@@ -273,6 +274,70 @@ func (cont *AciController) updateServicesForNode(nodename string) {
...
+					deleteTimeStr := device.GetAttrStr("deleteTime")
+					deleteTime, err := time.Parse(time.RFC3339, deleteTimeStr)
+					if err != nil {
+						cont.log.Error("Failed to parse opflex device delete time: ", err)
+						continue
+					}
+					now := time.Now()
+					diff := now.Sub(deleteTime)
+					if diff.Seconds() >= cont.config.DeviceDeleteTimeout {
+						devices = append(devices[:i], devices[i+1:]...)
```
Comment (snaiksat) [inline pkg/controller/services.go:309]: I am not sure I understand the logic here, aren't you iterating over "devices"? So, if L308 is true, it means you found a candidate device to delete? And if so, should you not make note of that device in a temporary list and set a flag "update" to True. L310 would not be required. Outside this for loop you check if the temporary list in not empty, and if so iterate through this temp list to delete the devices for this node (and if deleted add this node to the nodeUpdates list).

**Review by snaiksat — CHANGES_REQUESTED** [review]

File: pkg/controller/config.go [hunk 2]
```diff
@@ -209,6 +209,9 @@ type ControllerConfig struct {
 
 	// Enable creation of VmmInjectedLabel, default is false
 	EnableVmmInjectedLabels bool `json:"enable-vmm-injected-labels,omitempty"`
+
+	// Timeout to delete old opflex devices
+	DeviceDeleteTimeout float64 `json:"device-delete-timeout,omitempty"`
```
[the remark this answers is not in the document]
Comment (snaiksat) [reply pkg/controller/config.go:214]: per comment in acc-provision patch, please rename to opflex-device-delete-timeout

**Comment by snaiksat** [comment]
minor comment about renaming the config, other than that looks good...

File: pkg/controller/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (akhilamohanan) [reply pkg/controller/config.go]: addressed

**Comment by akhilamohanan** [comment]
Tested vm migration after applying the following patch to the PR:


diff --git a/pkg/controller/controller.go b/pkg/controller/controller.go
index 8659b9bdd..c7c0adce1 100644
--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -496,7 +496,7 @@ func (cont *AciController) Run(stopCh <-chan struct{}) {
 
        // If OpflexDeviceDeleteTimeout is not defined, default to 1800s
        if cont.config.OpflexDeviceDeleteTimeout == 0 {
-               cont.config.OpflexDeviceDeleteTimeout = 1800
+               cont.config.OpflexDeviceDeleteTimeout = 2
        }
 
        // If not defined, default to 32
diff --git a/pkg/controller/services.go b/pkg/controller/services.go
index 321c97e33..21e8da913 100644
--- a/pkg/controller/services.go
+++ b/pkg/controller/services.go
@@ -1112,6 +1112,7 @@ func (cont *AciController) postOpflexDeviceDelete(nodes []string) {
 }
 
 func (cont *AciController) opflexDeviceDeleted(dn string) {
+       return
        var nodeUpdates []string
        var dnFound bool //to check if the dn belongs to this cluster
        cont.indexMutex.Lock()

**Review by snaiksat — APPROVED** [review]

**Pull request merged**